src/math/src/number.rs

```diff
@@ -1,18 +1,523 @@
+    ///
+    /// ```
+    /// # use math::Number;
+    ///
+    /// # fn main() -> math::Result<()> {
+    ///     let random = Number::random();
+    ///     assert!(random.div(Number::ZERO).is_err());
+    ///     assert_eq!(Number::ZERO.div(random)?, Number::ZERO);
+    ///     assert_eq!(rand.div(rand)?, Number::ONE);
+    ///
+    ///     let a = Number::from(3);
+    ///     let b = Number::from(1.5);
+    ///     let neg_b = Number::from(-1.5);
+    ///     assert_eq!(a.div(b)?, Number::from(2));
+    ///     assert_eq!(a.div(neg_b)?, Number::from(-2));
+    /// #   Ok(())
+    /// # }
+    /// ```
+    pub fn div(&self, other: impl Into<Self>) -> Result<Self> {
+        todo!()
+    }
+
+    /// Raises self to the power of `exp`, using exponentiation by squaring.
+    ///
+    /// ```
+    /// # use math::Number;
+    /// assert_eq!(Number::random().pow(Number::ZERO), Ok(Number::ONE));
+    /// assert_eq!(Number::from(5).pow(2), Ok(Number::from(25)));
+    /// ```
+    pub fn power(&self, exp: impl Into<Self>) -> Result<Self> {
+        todo!()
+    }
+
+    /// Get the remainder of `self / other`
+    ///
+    /// ```
+    /// # use math::Number;
+    /// assert_eq!(Number::from(5).modulo(2), Ok(Number::ONE));
+    /// ```
+    pub fn modulo(&self, other: impl Into<Self>) -> Result<Self> {
+        todo!()
+    }
+
+    /// Get the absolute value of the given number
+    ///
+    /// ```
+    /// # use math::Number;
+    /// let a = Number::random();
+    /// let neg_a = a.mul(-1);
+    ///
+    /// assert_eq!(a.abs(), Ok(a));
+    /// assert_eq!(neg_a.abs(), Ok(a));
+    /// ```
+    pub fn abs(&self) -> Result<Self> {
+        todo!()
+    }
+
+    /// Calculate factorial of a given number
+    /// The number is not limited to integer, it can be a fraction
+    ///
+    /// # Error
+    /// return Error::FactorialNegative if the number is less than 0
+    ///
+    /// ```
+    /// # use math::Number;
+    /// assert_eq!(Number::Zero.factorial(), Number::One);
+    /// assert_eq!(Number::from(5).factorial(), Ok(Number::from(120)));
+    /// assert_eq!(Number::from(3.2).factorial().unwrap().to_string(6), "7.756689");
+    /// assert!(Number::from(-1).factorial().is_err());
+    /// ```
+    pub fn factorial(&self) -> Result<Self> {
+        todo!()
+    }
+
+    /// Returns the logarithm of the number with respect to an arbitrary `base`.
+    ///
+    /// # Error
+    /// Error::LogUndefinedBase if the `base` is less or equal than 0
+    /// Error::LogUndefinedNumber if the number is less or equal than 0
+    ///
+    /// ```
+    /// # use math::Number;
+    ///
+    /// # fn main() -> math::Result<()> {
+    ///     assert!(Number::random().log(0).is_err());
+    ///     assert!(Number::random().log(-1.2).is_err());
+    ///     assert!(Number::ZERO.log(Number::random()).is_err());
+    ///     assert!(Number::from(-0.3).log(Number::random()).is_err());
+    ///
+    ///     let a = Number::random();
```

Comment:
I am thinking, that using random floats in range <0, 1> in unit tests is not a good idea. Floating arithmetic can be weird, and thus the test could fail even when the computation is correct.

Reply:
Exactly that is the point of using random <0, 1>, the test should covers cases where things easily go wrong, [I'm currently looking into this issue with binary representation of a float point.](https://github.com/FIT-IVS-segmentation-fault-core-dumped/IVS-projekt-2/pull/5#discussion_r1155293072)